Fix bugzilla issue #18262 - Improve diagnostic for auto-increment on enums with enum base type

samrendra-singh opened the pull request:
Fixes bugzilla issue 18262.

**Overview**
This pull request improves the diagnostic message for auto-incrementing enum members when the base type is another enum. Previously, the compiler generated a confusing error message. Now, it clearly instructs the user to provide an explicit initializer.

**Problem**
When using an enum as the base type for another enum, the language specification requires that auto-increment is not performed because the base type does not support it. This patch checks the base type and emits a descriptive error message if auto-increment is attempted, which aligns with the intended behavior.

**Solution**
In enumsem.d, within the enumMemberSemantic function, I added a check to determine if the enum's base type is an enum. If it is, a clear error message is issued, and the auto-increment is aborted. This change prevents the confusing diagnostic previously encountered.

**Test**
I added a test case using a sample file that replicates the issue. The new error message is correctly displayed when compiling the file with the updated DMD.

TurkeyMan committed:
Accept __rvalue attribute on ref functions; which will force the result to be treated as __rvalue. (#20946)

This is essential to implement `move`, `forward`, etc.

Albert24GG committed:
Fix #17481 - free ctors before returning (#20975)

rikkimax committed:
Fix dub file for dscanner (#20978)

MatthewQiu-5 committed:
Extracted Condition.include to visitor in expressionsem (#20977)

WalterBright committed:
fix tyToExtend() (#20979)

thewilsonator committed:
fixup `mtype.h` for #20946 (#20971)

abulgit committed:
Fixes the invariant not called on extern(C++) classes issue (#20981)

WalterBright committed:
progress on dealing with common subexpressions (#20987)

WalterBright committed:
more common subexpression implementation (#20988)

abulgit committed:
Fix #20318 Compiler should explain why implicit conversion to mutable fails (#20984)

dkorpel committed:
Reset global Loc table in initDMD (#20995)

WalterBright committed:
get branches to block working (#20996)

* carith() function in exe1.c tests

* get branches to blocks working

WalterBright committed:
carith and more branch logic (#21002)

* carith() function in exe1.c tests

* more branch logic

dkorpel committed:
Improve Fix #20318 Compiler should explain why implicit conversion to mutable fails (#21003)

abulgit committed:
Fix #20075 - Improve error message for immutable constructor type mismatch (#20990)

WalterBright committed:
load sign extended bytes (#21004)

* carith() function in exe1.c tests

* load sign extended bytes

pbackus committed:
Don't parse lambda as UDA without parentheses (#21009)

This brings the compiler's behavior in line with the language spec.

Fixes dlang/dlang.org#4137

WalterBright committed:
it is IFL1 not IFL2 (#21010)

nybzmr committed:
Remove unused RTLSYM.NEWARRAYMTX (#21011)

WalterBright committed:
overlooked necessary sign extension (#21021)

abulgit committed:
Fix : Improved error message on override mismatch (#21023)

WalterBright committed:
flags not set by complement instruction (#21026)

shivangshukla7020 committed:
Refactor: Remove Unused Hooks in dinterpret.d (#21022)

- Removed references to _d_arrayappendcTX, _d_newThrowable, and _d_arrayappendT.
- Ensured test cases pass after deletion.
- Left _d_arrayctor and _d_arraysetctor untouched for future semantic phase changes.

dkorpel committed:
Move ImportC enum semantic to importc.d (#21027)

Co-authored-by: Dennis Korpel <dennis@sarc.nl>

nybzmr committed:
Remove Unused RTLSYMs from DMD (#21032)

* Remove RTLSYM for Translation PR #15819: Removed NEWARRAYMITX, NEWARRAYMITX, TRACENEWARRAYMTX and TRACENEWARRAYMITX

* Remove RTLSYM for Translation PR #15299: Removed NEWARRAYT, NEWARRAYIT, TRACENEWARRAYT and TRACENEWARRAYIT

* Remove RTLSYM for Translation PR #14837: Removed NEWCLASS, TRACENEWCLASS

* Remove RTLSYM for Translation PR #14664: Removed NEWITEMT, NEWITEMIT, TRACENEWITEMT and TRACENEWITEMIT

* Remove RTLSYM for Translation PR #14550: Removed ARRAYCATNTX, ARRAYCATT, TRACEARRAYCATNTX and TRACEARRAYCATT

* Remove RTLSYM for Translation PR #14382: Removed ARRAYSETASSIGN

* Remove RTLSYM for Translation PR #14310: Removed ARRAYASSIGN

* Remove RTLSYM for Translation PR #13495: Removed ARRAYAPPENDT, ARRAYAPPENDCTX, TRACEARRAYAPPENDT and TRACEARRAYAPPENDCTX

samrendra-singh committed:
Fix issue 18262 - Improve diagnostic for auto-increment on enums with enum base type

When an enum member is auto-incremented and the enum’s base type is another enum,
the current behavior produces a confusing error message. This patch adds a check
to detect when the base type is an enum and produces a clearer, more descriptive
error message instructing the user to provide an explicit initializer.

samrendra-singh committed:
Fix issue 18262 - Improve diagnostic for auto-increment on enums with enum base type

When an enum member is auto-incremented and the enum’s base type is another enum,
the current behavior produces a confusing error message. This patch adds a check
to detect when the base type is an enum and produces a clearer, more descriptive
error message instructing the user to provide an explicit initializer.

Add fail_compilation test for enum auto-increment restriction (Bug 18262)

samrendra-singh committed:
Merge branch 'fix-18262-autoIncrementError' of https://github.com/solo-source/dmd into fix-18262-autoIncrementError

samrendra-singh committed:
Fix Bugzilla Issue 18262 - Improve diagnostic for auto-increment on enums with enum base type

When an enum member is auto-incremented and the enum’s base type is another enum,
the current behavior produces a confusing error message. This patch adds a check
to detect when the base type is an enum and produces a clearer, more descriptive
error message instructing the user to provide an explicit initializer.

samrendra-singh committed:
Fix Bugzilla Issue 24684 - Improve diagnostic for auto-increment on enums with enum base type.

When an enum member is auto-incremented and the enum’s base type is another enum,
the current behavior produces a confusing error message. This patch adds a check
to detect when the base type is an enum and produces a clearer, more descriptive
error message instructing the user to provide an explicit initializer.

samrendra-singh committed:
Merge branch 'fix-18262-autoIncrementError' of https://github.com/solo-source/dmd into fix-18262-autoIncrementError

samrendra-singh committed:
Fix Issue#18262 - Improve diagnostic for auto-increment on enums with enum base type

When an enum member is auto-incremented and the enum’s base type is another enum,
the current behavior produces a confusing error message. This patch adds a check
to detect when the base type is an enum and produces a clearer, more descriptive
error message instructing the user to provide an explicit initializer.

samrendra-singh committed:
Merge branch 'fix-18262-autoIncrementError' of https://github.com/solo-source/dmd into fix-18262-autoIncrementError

solo-source commented:
The mention of Bugzilla Issue 18262 is by pure accident, i am only trying to fix Bugzilla Issue 24684

samrendra-singh committed:
Merge branch 'fix-18262-autoIncrementError' of https://github.com/solo-source/dmd into fix-18262-autoIncrementError

samrendra-singh committed:
Fix Issue#18262 - Trying to fix the issues

samrendra-singh committed:
Fix Issue#18262 - Improve diagnostic for auto-increment on enums with enum base type

When an enum member is auto-incremented and the enum’s base type is another enum,
the current behavior produces a confusing error message. This patch adds a check
to detect when the base type is an enum and produces a clearer, more descriptive
error message instructing the user to provide an explicit initializer.

solo-source commented:
I am new to Creating PR requests, thats why all this mess happened. Please Ignore the mess, i think the latest commit is what i wanted to do.

 I was just trying to edit the commit message so that the issue can be referenced directly, but i dont think that is happening. Anyways i apologise for this mess.

samrendra-singh committed:
Merge branch 'dlang:master' into fix-18262-autoIncrementError

solo-source commented:
i messed up this PR very badly, so i am closing it now.
